### PR TITLE
improve lowering to LLVM and update submodule

### PR DESCRIPTION
**Pull request by yhls**

### CODE REVIEW

File: lib/Conversion/JLIRToLLVM/JLIRToLLVM.cpp [hunk 1]
```diff
@@ -116,52 +184,105 @@ struct FuncOpConversion : public OpConversionPattern<FuncOp> {
...
+        TypeConverter::SignatureConversion result(op.getNumArguments());
         for (auto &en : llvm::enumerate(type.getInputs())) {
-            LLVM::LLVMType converted = lowering.convertToLLVMType(en.value());
+            LLVM::LLVMType converted =
+                lowering.convertToLLVMType(en.value());
             assert(converted && "failed to convert argument type");
-            result.addInputs(en.index(), converted);
+
+            // drop argument if it converts to void type
+            if (converted == lowering.void_type) {
```
[the remark this answers is not in the document]
Comment (yhls) [reply lib/Conversion/JLIRToLLVM/JLIRToLLVM.cpp]: Hmm, I think these arguments could be useful to keep around while we are still in JLIR, so it makes sense to drop them when or after converting to the LLVM dialect, but the LLVM dialect doesn't like arguments of void type.

File: lib/Conversion/JLIRToLLVM/JLIRToLLVM.cpp [hunk 2]
```diff
@@ -93,16 +132,45 @@ struct ToUndefPattern : public OpConversionPattern<SourceOp> {
...
-struct FuncOpConversion : public OpConversionPattern<FuncOp> {
-    JLIRToLLVMTypeConverter &lowering;
+template <typename SourceOp, typename CmpOp, typename Predicate, Predicate predicate>
+struct ToCmpOpPattern : public OpAndTypeConversionPattern<SourceOp> {
+    using OpAndTypeConversionPattern<SourceOp>::OpAndTypeConversionPattern;
+
+    PatternMatchResult matchAndRewrite(SourceOp op,
+                                       ArrayRef<Value> operands,
+                                       ConversionPatternRewriter &rewriter) const override {
+        assert(operands.size() == 2);
```
[the remark this answers is not in the document]
Comment (yhls) [reply lib/Conversion/JLIRToLLVM/JLIRToLLVM.cpp]: Good point, I've put writing verifiers on my list

File: include/brutus/Dialect/Julia/JuliaOps.h [hunk 3]
```diff
@@ -78,6 +78,15 @@ class JuliaType
     static bool kindof(unsigned kind) { return kind == JLIRTypes::JuliaType; }
 
     static JuliaType get(mlir::MLIRContext *context, jl_datatype_t *datatype) {
+        jl_module_t *core_module = (jl_module_t*)jl_get_global(jl_main_module, jl_symbol("Core"));
```
[the remark this answers is not in the document]
Comment (yhls) [reply include/brutus/Dialect/Julia/JuliaOps.h]: Fixed

File: lib/Codegen/Brutus.cpp [hunk 4]
```diff
@@ -1,18 +1,24 @@
 #include "brutus/brutus.h"
+#include "brutus/brutus_internal.h"
 #include "brutus/Dialect/Julia/JuliaOps.h"
+
 #include "mlir/InitAllDialects.h"
 
 extern "C" {
-jl_sym_t *call_sym;    jl_sym_t *invoke_sym;
```
Comment (vchuravy) [inline lib/Codegen/Brutus.cpp]: I thought this is necessary to anchor the extern variables in this compilation unit

Comment (yhls) [reply lib/Codegen/Brutus.cpp]: Oops, you're right, they still need to be defined somewhere

**Pull request merged**